packages/vite/src/node/server/index.ts

```diff
@@ -115,6 +115,10 @@ export interface ServerOptions {
    * Create Vite dev server to be used as a middleware in an existing server
    */
   middlewareMode?: boolean
+  /**
+   * Serve and fallback index.html in middleware mode
+   */
+  middlewareModeServeHtml?: boolean
```

Comment:
**thought:** @patak-js What do you think about using `middlewareMode?: boolean | MiddlewareModeOptions` instead?
So we would have it like `cors` that can be set to `true` or can have configured more options if needed.

Reply:
Yes, I think the options could be `boolean | { spaFallback, transformIndexHtml }` to start with so you can choose the middleware you want to enable. I think the default should be true for both, so `{}` means use middleware mode with everything enabled, and you can disable the ones you don't want to use.

Reply:
> Yes, I think the options could be `boolean | { spaFallback, transformIndexHtml }` to start with so you can choose the middleware you want to enable. I think the default should be true for both, so `{}` means use middleware mode with everything enabled, and you can disable the ones you don't want to use.

Why not everything `false`? At present if we set `middlewareMode` to `true`, `spaFallback` and `transformIndexHtml` are both disabled.
Should we also make them default to `true` when `middlewareMode` is `true`? That will change the behavior.

Reply:
We shouldn't change the behavior of `middlewareMode: true`. The defaults there are good for the common case. 

When you opt-in to selecting what should be enabled/disabled `{}`, we could:
a. make it work in the same way as `true`.
b. everything `false`
c. everything `true`

All of them have pros and cons. And I am starting to think that we are going to be exposing too much of the internal middlewares with this, and we may want to avoid this to be able to change them later.

I am thinking that it may be better to offer:
```
middlewareMode: 'html' | 'ssr' | true
```
Where `true` is the same as `'ssr'`. `'html'` may also be called `'all'` or `'enabled'`. I think we can prepare the PR with this approach and ask Evan what is his opinion later.

Reply:
> We shouldn't change the behavior of `middlewareMode: true`. The defaults there are good for the common case.
> 
> When you opt-in to selecting what should be enabled/disabled `{}`, we could:
> a. make it work in the same way as `true`.
> b. everything `false`
> c. everything `true`
> 
> All of them have pros and cons. And I am starting to think that we are going to be exposing too much of the internal middlewares with this, and we may want to avoid this to be able to change them later.
> 
> I am thinking that it may be better to offer:
> 
> ```
> middlewareMode: 'html' | 'ssr' | true
> ```
> 
> Where `true` is the same as `'ssr'`. `'html'` may also be called `'all'` or `'enabled'`. I think we can prepare the PR with this approach and ask Evan what is his opinion later.

Some string values like `'all'` and `'enabled'` may be a little confusing (all what?).

Because both `spaFallback` and `indexHtmlMiddleware` relate to html serving, so I think we could have an object options which has only one property:

```
MiddlewareMode {
  // use the built-in html serving for development
  // default to false
  serveHtml: boolean;
}
```